[addons] add parts to CAddonInfo to read addon.xml

AlwinEsch opened the pull request:
## Description
<!--- Describe your change in detail -->
This change is the first big major part to remove cpluff on next
steps.

On them here becomes the parts for addon info (old props) readed by
CAddonInfo class contructor, by give of path to addon or with a tinyXML part
(used later for repo parts). With them is cpluff no more needed to read
addon.xml, further is it direct designed and required to support multiple
types (e.g. PVR with inputstream or inputstream with videocodec).

Screensaver addon is with them changed to create with new way.

Temporary is for this on CAddonBuilder the from cpluff given path used for
creation of CAddonInfo.

On next steps becomes all addon types changed to this way and shortly only
the from cpluff given parts used.

<!--- Provide a general summary of your change in the Title above -->


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/addons/AddonInfo.h:
```diff
@@ -81,16 +84,128 @@ namespace ADDON
+  typedef std::vector<std::pair<std::string, extValue> > EXT_VALUE;
+  typedef std::vector<std::pair<std::string, EXT_VALUE> > EXT_VALUES;
+
+  class CAddonBuilder;
```
Is this forward decl actually needed?

AlwinEsch replied:
Is for the temporary declared as child class below

ksooo commented on xbmc/addons/AddonInfo.h:
```diff
@@ -81,16 +84,128 @@ namespace ADDON
+    EXT_ELEMENTS m_childs;
+  };
+
+  class CAddonInfo;
```
Is this forward decl actually needed?

AlwinEsch replied:
Is for the contructor on CAddonType and also is CAddonType used from CAddonInfo, thats why the decl was needed.

ksooo commented on xbmc/addons/AddonInfo.h:
```diff
@@ -81,16 +84,128 @@ namespace ADDON
+
+  private:
+    friend class ADDON::CAddonBuilder; // temporary until cpluff is removed
+    friend class CAddonInfo;
```
Can't this be avoided?

AlwinEsch replied:
Yes, was only a test mistake from.

ksooo commented on xbmc/addons/AddonInfo.h:
```diff
@@ -107,13 +222,18 @@ namespace ADDON
     const std::string& Summary() const { return m_summary; }
     const std::string& Description() const { return m_description; }
-    const std::string& LibName() const { return m_libname; }
+    const std::string& LibName() const { return m_types[0].m_libname; }
```
Again, this looks strange. Why is it always the libname for the first type?

AlwinEsch replied:
Here it is needed in future for binary addons. I think to change the name to `MainLibName()`.

ksooo reviewed (changes requested):
A couple of questions/suggestions...

AlwinEsch commented on xbmc/addons/Addon.cpp:
```diff
@@ -59,6 +59,17 @@ using XFILE::CFile;
 
+  CAddon::CAddon(const std::string& addonPath)
+  : m_addonInfo(addonPath)
+  , m_userSettingsPath()
```
c&p becomes removed.

AlwinEsch commented on xbmc/addons/AddonBuilder.h:
```diff
@@ -45,7 +45,7 @@ class CAddonBuilder
   void SetBroken(std::string broken) { m_addonInfo.m_broken = std::move(broken); }
   void SetPath(std::string path) { m_addonInfo.m_path = std::move(path); }
-  void SetLibName(std::string libname) { m_addonInfo.m_libname = std::move(libname); }
+  void SetLibName(std::string libname) { m_addonInfo.m_types[0].m_libname = std::move(libname); }
```
This is a temporary for old cpluff where always only one lib name is. On new way can be this no more direct accessed.

For a binary can it be only one, but on Python is it mostly on different scripts.

AlwinEsch commented on xbmc/addons/AddonInfo.cpp:
```diff
@@ -77,6 +83,207 @@ static const TypeMapping types[] =
+{
+  const char* cstring;
+
+  cstring = element->Attribute("point");
```
Can change, it becomes set also from another places thats why I leaved alone.

AlwinEsch commented on xbmc/addons/AddonInfo.cpp:
```diff
@@ -77,6 +83,207 @@ static const TypeMapping types[] =
+
+const CAddonExtensions* CAddonExtensions::GetElement(const std::string& id) const
+{
+  for (EXT_ELEMENTS::const_iterator it = m_childs.begin(); it != m_childs.end(); ++it)
```
Is better, still have mostly old style in mind ;-)

tamland commented:
Getting rid of the cpluff parser is great!, so sorry to be blunt here, but design wise this turning into a huge mess. Everything you are trying to do here have already been solved a long time ago!

What you need to do instead is:

1. Add a new factory method similar to [this](https://github.com/xbmc/xbmc/blob/a0a04d6011eedfd149709669e3526cd5cd1abce1/xbmc/addons/AddonManager.h#L254) that uses tinyXML instead of cpluff. (Or even better: make a new factory class)

2. Move all xml parsing logic from CAddonInfo and CAddonType constuctors to here. 

3. You then pass the parsed info, including a fully constructed vector<CAddonType> to the CAddonInfo constructor

tamland commented on xbmc/addons/AddonInfo.cpp:
```diff
@@ -151,7 +414,448 @@ CAddonInfo::CAddonInfo(std::string id, TYPE type)
+  return contents > 0 ? true : false;
+}
+
+std::string CAddonInfo::SerializeMetadata()
```
This is a database implementation detail and is tied to the schema and migration. This should *not* be moved into to a generic class, but kept as a private function in CAddonDatabse.

AlwinEsch replied:
This is a Chicken and egg problem,

CAddonInfo should hold everything private to allow only a read of the values but no changes.
Reason behind is that the class is on end only created one time and hold in a list from AddonManager. If some part need then something, becomes the already stored CAddonInfo from the list used.

Only way but also not the best is to add the database as "friend" to CAddonInfo.

See also the coming parts on other comment.

tamland replied:
Yet, in the current code all values are read only, this function is in CAddonDatabse, and there are no friend classes needed. 

If you need to break that to make it "fit" your solution, then that solution is the wrong one.

AlwinEsch replied:
It is a from database used part complete right!

But I must say again, to have in Database is complete bad!

It use values from CAddonInfo and use only a string as "in" or "out" for database. All used values from the string are from CAddonInfo.

**In case now new values becomes added (removed bring God thanks to a compile error), is the database be easily forgotten!**

Further is CAddonInfo not a generic class, it is the base addon class (and on end only one time for one addon present) to declare the whole addon in Kodi.

tamland replied:
The point here is having some basic separation of concern. CAddonDatabse is responsible for mapping an unmapping. By exposing the raw unmapping `metadata` value, CAddonInfo now has to details about how things are stored in the database.

I don't see you put reading/writing from the db `Dataset` in this class so why do it with `metadata`? (hint: you don't because mapping of `Dataset` is the entire job of the database). The `metadata` value is the same thing as `Dataset`, it just happens to be json right now.

tamland replied:
And here's a lengthy explanation for why it's done this way: https://msdn.microsoft.com/en-us/library/ff649690.aspx

FernetMenta replied:
even Microsoft architects call this an anti pattern: http://www.infoworld.com/article/3117713/application-development/design-patterns-that-i-often-avoid-repository-pattern.html

tamland replied:
@FernetMenta you should read the link you post.. He call the *generic* repository pattern (over entity frameworks) an anti-pattern. And I agree.

tamland commented on xbmc/addons/AddonInfo.cpp:
```diff
@@ -151,7 +414,448 @@ CAddonInfo::CAddonInfo(std::string id, TYPE type)
+
+      m_types.push_back(CAddonType(type, this, child));
+    }
+  }
```
There are multiple errors in this method compared to original cpluff implementation. They are easily found by running it again the existing tests: https://github.com/xbmc/xbmc/blob/a0a04d6011eedfd149709669e3526cd5cd1abce1/xbmc/addons/test/TestAddonFactory.cpp

AlwinEsch replied:
The existing tests must be complete reworked after cpluff is removed. Everything related on current test parts must be ignored for this.

If you see a error please describe me detailed what it bring.

AlwinEsch commented:
Hi @tamland, to remove the cpluff is not in a easy small step possible.

To change the complete addon system to new way is a big part.

Currently is this CAddonInfo still as static in CAddon class. This is must be removed complete (on next requests) and only given as shared_ptr to the addon class.

Goal is (and used on kodi-agile) that Kodi e.g. go over the addon directory, if there is a addon with addon.xml becomes this by the constructor of CAddonInfo read and the class stored in a list on addon manager. What Kodi does with it does not interest. This mean also a related addon class is stll not needed and created.

After all steps are done contains the addon class only a `CAddon(std::shared_ptr<CAddonInfo> addonInfo)` contructor where the stored parts becomes given.

XML parsing must only be done inside this class and should not done somewhere outside. To do outside bring a lot bad parts and nothing good.

Big problem is to bring in without destroy something and allow a better review. This need for short times bad parts. Thats why I mostly use screensaver where is not so critical.

tamland commented:
> XML parsing must only be done inside this class and should not done somewhere outside. To do outside bring a lot bad parts and nothing good.

Sorry but this is a horrible idea. There nothing 'bad' about keeping it outside. It's called the factory pattern and it's great for this sort of thing. Not too long ago this exact 'inside this class' spaghetti was removed. And the reasons for that was:

- Having complex logic like parsing and filesystem code in constructor is an anti-pattern. 

- It breaks testability.

- It makes error handing complicated as your only option is to leave objects in an invalid state and work around it by setting flags (this is exactly what you have done with m_usable)

- It's a pain to deal with configuration as every new config must be added to every constructors. 

- Makes it a lot harder to construct something partially when needed. E.g. local addons have metadata from the filesystem and from the database, and potentially also from the repository and they must all be combined.

Additionally, the way this is implemented violates the single responsibility principle. `CAddonInfo` now holds the info, it holds state, it's an xml parser, it's an json parser, it's responsible for the extensions. 
The same for `CAddonType`: it's hold extension info, it's a parser, it's a mutable collection. This is all effects of having the classes themselves do the parsing.

The solution I explained in my previous post does not have these issues.

AlwinEsch commented:
Maybe I improve again

AlwinEsch commented:
Excuse me, but after over two years I want and I have to come to an end.

Always comes a no and see how the current addon interface is getting more and more mixed up or hacking paths to achieve a goal.

Since I am surprised again and again something larger is rejected by me but otherwise which is unreliable somewhere else comes in.

Excuse me that I bring these words, we are a good team.

But I do not want again this cross and cross between the different parts, this brings again and again problems since the one then no longer fits with the other.

So far, where the AddonProps was recreated with every little thing, it was easy, but since in the future this is done only once must be ensured that does not change from now on equal in what.

**I look at what I can do, but I want to keep a few things from it.**

Sorry for my words

AlwinEsch commented:
I create now a another request to remove the static declaration of CAddonInfo (AddonProps) in CAddon and pass as shared pointer to them.

Is a part also required for them here.

tamland commented:
@AlwinEsch You can't just throw out all the existing code and design choices, make a complete re-implementation with tons of issues, and then expect everything to be just fine. (That's not collaboration).

I have tried to tell you this and propose better solutions multiple times, in this pr and others, but you seem too fixated on continuing down path you have already decided on, ignoring any suggestions.

Why don't you just *try* and do what [I proposed earlier](https://github.com/xbmc/xbmc/pull/12283#issuecomment-307575101)? Everything you want to achieve with removing cpluff and supporting multiple types fits well within the current design with minor changes. I know it does because everything here has already been made with this in mind long before you started with it.

ksooo commented:
For me, what @tamland suggested makes sense. I especially would be happy if he's right on that not everything needs to be reimplemented because the existing design already fits well an can be extended with not too much effort.

ksooo commented:
Maybe you two guys should discuss this 1:1 on Slack?

AlwinEsch commented:
I write you something on slack.

But I say, **not** expect everything to be just fine, never possible on world. But it must come a day where the old way becomes changed to a new.

AlwinEsch commented:
Is updated once, thanks @ksooo for the comments.

But not yet final.

AlwinEsch committed:
[addons] add parts to CAddonInfo to read addon.xml

This change is the first big major part to remove cpluff on next
steps.

On them here becomes the parts for addon info (old props) readed by
CAddonInfo class contructor, by give of path to addon or with a tinyXML part
(used later for repo parts). With them is cpluff no more needed to read
addon.xml, further is it direct designed and required to support multiple
types (e.g. PVR with inputstream or inputstream with videocodec).

On next steps becomes all addon types changed to this way and shortly only
the from cpluff given parts used.

ksooo commented on xbmc/addons/AddonInfo.cpp:
```diff
@@ -142,7 +352,63 @@ CAddonInfo::CAddonInfo()
+    m_mainType(ADDON_UNKNOWN),
+    m_packageSize(0)
+{
+  m_usable = LoadAddonXML(baseElement, URIUtils::AddFileToFolder(addonRepoXmlPath, m_id));
```
This could be moved to init list.

ksooo commented on xbmc/addons/AddonInfo.cpp:
```diff
@@ -142,7 +352,63 @@ CAddonInfo::CAddonInfo()
+    m_origin(origin),
+    m_packageSize(0)
+{
+  m_usable = DeserializeMetadata(metadata);
```
This could be moved to init list.
